### PR TITLE
Adding java opts to config

### PR DESCRIPTION
**Pull request by franciscolopezsancho**
### What changes were proposed in this pull request?
Being able to pass JAVA_OPTS as env in a flink streamlet 
as per https://github.com/lightbend/cloudflow/issues/624

### Why are the changes needed?
to allow to add env.java.opts in a Flink streamlet in a consistent manner respect to the rest of the streamlets

### Does this PR introduce any user-facing change?
yes, as the possibility of passing that variable

### How was this patch tested?
through the test `"read values from pod configuration in JAVA_OPTS and put it in Flink conf " +
      "as env.java.opts" in `
in `FlinkRunnerSpec`

### CODE REVIEW

File: core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala [hunk 1]
```diff
@@ -78,6 +78,84 @@ class FlinkRunnerSpec extends WordSpecLike with OptionValues with MustMatchers w
       replicas = None
     )
 
+    "read values from pod configuration and put it in both " in {
```
Comment (olofwalker) [inline core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala:81]: "...taskmanager and jobmanager" ?

File: core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala [hunk 2]
```diff
@@ -78,6 +78,84 @@ class FlinkRunnerSpec extends WordSpecLike with OptionValues with MustMatchers w
...
+      )
+      crd.spec.jobManagerConfig.envConfig.get.env.get mustBe Vector(EnvVar("FOO",EnvVar.StringValue("BAR")))
+      crd.spec.taskManagerConfig.envConfig.get.env.get mustBe Vector(EnvVar("FOO",EnvVar.StringValue("BAR")))
+      crd.spec.jobManagerConfig.resources.get.requests mustBe Map(Resource.memory -> "512M")
+      crd.spec.taskManagerConfig.resources.get.requests mustBe Map(Resource.memory -> "512M")
+      crd.spec.jobManagerConfig.resources.get.limits mustBe Map(Resource.memory -> "1024M")
+      crd.spec.taskManagerConfig.resources.get.limits mustBe Map(Resource.memory -> "1024M")
+    }
+
+    "read values from pod configuration in JAVA_OPTS and put it in Flink conf " +
```
Comment (olofwalker) [inline core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala:133]: Is the reason for this concatenation to keep the line length down?

**Comment by olofwalker** [comment]
@franciscolopezsancho I think you need to run a `scalafmtAll` also

File: core/cloudflow-operator/src/main/scala/cloudflow/operator/runner/FlinkRunner.scala [hunk 3]
```diff
@@ -44,7 +44,6 @@ object FlinkRunner extends Runner[CR] {
   final val runtime         = "flink"
   final val PVCMountPath    = "/mnt/flink/storage"
   final val DefaultReplicas = 2
-  final val JvmArgsEnvVar   = "JVM_ARGS"
```
Comment (RayRoestenburg) [inline core/cloudflow-operator/src/main/scala/cloudflow/operator/runner/FlinkRunner.scala:47]: 👍

File: core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala [hunk 4]
```diff
@@ -78,6 +78,115 @@ class FlinkRunnerSpec extends WordSpecLike with OptionValues with MustMatchers w
       replicas = None
     )
 
+    "read values from pod configuration and put it in both " in {
```
Comment (RayRoestenburg) [inline core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala:81]:
```suggestion
    "configure the FlinkApp from JAVA_OPTS in kubernetes container env as well as from flink config env.java.opts" in {
```

File: core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala [hunk 5]
```diff
@@ -78,6 +78,115 @@ class FlinkRunnerSpec extends WordSpecLike with OptionValues with MustMatchers w
...
+      crd.spec.image must include(imageWithoutRegistry)
+      crd.spec.entryClass mustBe "cloudflow.runner.Runner"
+
+      crd.spec.volumes mustBe Vector(
+        Volume("config-map-vol", Volume.ConfigMapVolumeSource("configmap-some-app-id")),
+        Volume("persistent-storage-vol", Volume.PersistentVolumeClaimRef("some-app-id-pvc")),
+        Volume("secret-vol", Volume.Secret("flink-streamlet")),
+        Runner.DownwardApiVolume
+      )
+      crd.spec.jobManagerConfig.envConfig.get.env.get mustBe Vector(EnvVar("FOO",EnvVar.StringValue("BAR")))
```
Comment (RayRoestenburg) [inline core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala:125]: You must add a check for the JAVA OPTS getting set here.

File: core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala [hunk 6]
```diff
@@ -78,6 +78,115 @@ class FlinkRunnerSpec extends WordSpecLike with OptionValues with MustMatchers w
...
+      crd.spec.taskManagerConfig.envConfig.get.env.get mustBe Vector(EnvVar("FOO",EnvVar.StringValue("BAR")))
+      crd.spec.jobManagerConfig.resources.get.requests mustBe Map(Resource.memory -> "512M")
+      crd.spec.taskManagerConfig.resources.get.requests mustBe Map(Resource.memory -> "512M")
+      crd.spec.jobManagerConfig.resources.get.limits mustBe Map(Resource.memory -> "1024M")
+      crd.spec.taskManagerConfig.resources.get.limits mustBe Map(Resource.memory -> "1024M")
+    }
+
+    "read values from pod configuration key JAVA_OPTS and " +
```
Comment (RayRoestenburg) [inline core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala:133]: Oh wait...

File: core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala (hunk 4 above)
Comment (RayRoestenburg) [inline core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala:81]:
```suggestion
    "read values from pod configuration and put it in both " in {
```
```suggestion
    "read environment variables and resource requirements from pod configuration and configure taskmanager and jobmanager " in {
```

File: core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala [hunk 7]
```diff
@@ -78,6 +78,115 @@ class FlinkRunnerSpec extends WordSpecLike with OptionValues with MustMatchers w
...
+      "put it in Flink conf in env.java.opts" in {
```
Comment (RayRoestenburg) [inline core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala:134]: no need to break line I think or is this scalafmt?

File: core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala [hunk 8]
```diff
@@ -78,6 +78,115 @@ class FlinkRunnerSpec extends WordSpecLike with OptionValues with MustMatchers w
...
+              |          { name = "JAVA_OPTS"
+              |            value = "-XX:MaxRAMPercentage=40.0"
+              |          }
+              |        ]
+              |}
+              |        """.stripMargin.getBytes(),
+            cloudflow.operator.event.ConfigInputChangeEvent.RuntimeConfigDataKey ->
+              """
+                |flink {
+                |            env.java.opts = "-XX:-DisableExplicitGC"
```
Comment (RayRoestenburg) [inline core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala:179]: Maybe format with some less indentation.. very minor

File: core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala [hunk 9]
```diff
@@ -78,6 +78,115 @@ class FlinkRunnerSpec extends WordSpecLike with OptionValues with MustMatchers w
...
+              |}
+              |        """.stripMargin.getBytes())
+        ),
+        namespace = namespace
+      )
+
+      crd.spec.flinkConfig.get("env.java.opts") mustBe Some("-XX:MaxRAMPercentage=40.0")
+    }
+
+    "read env.java.opts from runtime Flink conf should override pod JAVA_OPTS value" in {
```
Comment (RayRoestenburg) [inline core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala:158]:
```suggestion
    "configure env.java.opts from runtime Flink conf, overriding what is provided as JAVA_OPTS value in the pod configuration" in {
```

File: core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala (hunk 1 above)
Comment (franciscolopezsancho) [reply core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala]: done

File: core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala (hunk 7 above)
Comment (franciscolopezsancho) [reply core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala]: my bad

File: core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala (hunk 2 above)
Comment (franciscolopezsancho) [reply core/cloudflow-operator/src/test/scala/cloudflow/operator/runner/FlinkRunnerSpec.scala]: yep it was, not anymore though

**Comment by franciscolopezsancho** [comment]
Has been tested on a cluster as well

**Pull request merged**